### PR TITLE
Improve SSO auth flow

### PR DESCRIPTION
**Pull request by t3chguy**
Requires https://github.com/vector-im/element-web/pull/16292
Fixes https://github.com/vector-im/element-web/issues/16219

+ Use replaceState instead of a redirect to strip the loginToken
+ Put user into the same post-auth flows of E2ESetup
+ Skip UIA prompt in this post-auth flow, happy path is a server grace period
+ Respect SSO IdP brand styling
+ Improve SSO UIA flow by: 
  + auto-closing the tab
  + better error on failure

Reviewing commit-by-commit is recommended.

image

### CODE REVIEW

File: res/img/element-icons/brands/apple.svg [hunk 1]
```diff
@@ -0,0 +1,3 @@
+<svg width="25" height="24" viewBox="0 0 25 24" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (jryans) [inline res/img/element-icons/brands/apple.svg:1]: (Off-topic) I'm not really sure why we have an `element-icons` directory at all... But anyway, it's been around for a while.

File: src/components/structures/InteractiveAuth.js [hunk 2]
```diff
@@ -177,7 +177,13 @@ export default class InteractiveAuthComponent extends React.Component {
             stageState: stageState,
             errorText: stageState.error,
         }, () => {
-            if (oldStage != stageType) this._setFocus();
+            if (oldStage !== stageType) {
+                this._setFocus();
+            } else if (!stageState.error && this._stageComponent.current &&
+                this._stageComponent.current.attemptFailed
+            ) {
```
Comment (jryans) [inline src/components/structures/InteractiveAuth.js:184]: Hmm, I find these multi-line conditionals a bit more readable when all lines start at the same column:

```suggestion
            } else if (
                !stageState.error && this._stageComponent.current &&
                this._stageComponent.current.attemptFailed
            ) {
```

File: src/components/views/auth/InteractiveAuthEntryComponents.js [hunk 3]
```diff
@@ -609,21 +609,46 @@ export class SSOAuthEntry extends React.Component {
...
     }
 
     componentDidMount(): void {
         this.props.onPhaseChange(SSOAuthEntry.PHASE_PREAUTH);
     }
 
+    componentWillUnmount() {
+        window.removeEventListener("message", this._onReceiveMessage);
+        if (this._popupWindow) {
+            this._popupWindow.close();
```
Comment (jryans) [inline src/components/views/auth/InteractiveAuthEntryComponents.js:628]:
```suggestion
            this._popupWindow.close();
            this._popupWindow = null;
```

File: src/components/views/auth/InteractiveAuthEntryComponents.js [hunk 4]
```diff
@@ -609,21 +609,46 @@ export class SSOAuthEntry extends React.Component {
...
+    attemptFailed = () => {
+        this.setState({
+            attemptFailed: true,
+        });
+    };
+
+    _onReceiveMessage = event => {
+        if (event.data === "authDone" && event.origin === this.props.matrixClient.getHomeserverUrl()) {
+            if (this._popupWindow) {
+                this._popupWindow.close();
```
Comment (jryans) [inline src/components/views/auth/InteractiveAuthEntryComponents.js:641]:
```suggestion
                this._popupWindow.close();
                this._popupWindow = null;
```

**Review by jryans — APPROVED** [review]
Thanks, looks like a great improvement! 😄

File: res/img/element-icons/brands/apple.svg (hunk 1 above)
Comment (t3chguy) [reply res/img/element-icons/brands/apple.svg:1]: I think it was for when the rebrand replaced a bunch but not all of the icons

File: res/css/views/elements/_SSOButtons.scss [hunk 5]
```diff
@@ -64,11 +78,6 @@ limitations under the License.
     border-color: #3c5a99;
     color: #ffffff;
 }
-.mx_SSOButton.mx_SSOButton_brand_google {
```
Comment (jryans) [inline res/css/views/elements/_SSOButtons.scss]: Hmm, no colours needed for Google now?

**Review by jryans — APPROVED** [review]
Feel free to merge if ready!

File: res/css/views/elements/_SSOButtons.scss (hunk 5 above)
Comment (t3chguy) [reply res/css/views/elements/_SSOButtons.scss]: Indeed, the branding guidelines show it as a colourful logo on a white button. The old figma for it was white logo on red button but thats Google+

Comment (t3chguy) [reply res/css/views/elements/_SSOButtons.scss]: image

**Pull request merged**